### PR TITLE
Defer shard pruning for fast-path router queries to execution

### PR DESCRIPTION
**Pull request by onderkalaci**
DESCRIPTION: Defer shard pruning for fast-path router queries to execution

This is purely to enable better performance with prepared statements.
Before this commit, the fast path queries with prepared statements
where the distribution key includes a parameter always went through
distributed planning. After this change, we only go through distributed
planning on the first 5 executions.

**In a CPU bottlenecked scenario, this PR helps to achieve %3-%4 more operations.**

The edge cases I've hit, and planning to add some more tests:

- [x] When SET citus.enable_fast_path_router_planner TO false; there are some edge cases. I think I've covered them all, still will look

- [x] Tests around WHERE false: ` prepare p1(int) as update users_table set value_3 = 0 where   user_id = $1 and false;` or `prepare p2(int) as select count(*) from  users_table  where   user_id = $1 and false;`

- [ ] Understand why the cursor tests has changed

- [x] Do we really need the `TaskListContainsInvalidAnchorShardId ()` check?

- [x] Move more code into common functions between router planner vs CitusBeginScan

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -1673,13 +1669,34 @@ RouterJob(Query *originalQuery, PlannerRestrictionContext *plannerRestrictionCon
...
 	bool requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
+	FastPathRestrictionContext *fastPathRestrictionContext =
+		plannerRestrictionContext->fastPathRestrictionContext;
+
+	/*
+	 * We prefer to defer shard pruning/task generation to the
+	 * execution when the parameter on the distribution key
+	 * cannot be resolved (or no paramater exists).
+	 */
+	if (fastPathRestrictionContext->fastPathRouterQuery &&
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c]: One idea I had was to make this decision more explicit by adding  `fastPathRestrictionContext -> distKeyEqualsParam` , and checking for it here.

Any preference?

File: src/test/regress/expected/fast_path_router_modify.out [hunk 2]
```diff
@@ -207,19 +207,19 @@ ERROR:  could not run distributed query with FOR UPDATE/SHARE commands
 HINT:  Consider using an equality filter on the distributed table's partition column.
 -- very simple queries on reference tables goes through fast-path planning
 DELETE FROM modify_fast_path_reference WHERE key = 1;
-DEBUG:  Distributed planning for a fast-path router query
+DEBUG:  Deferred pruning for a fast-path router query
```
Comment (onderkalaci) [inline src/test/regress/expected/fast_path_router_modify.out]: For example, we defer pruning here just because it is a reference table and we cannot find a distribution key value.

There are some other examples for deferring pruning just because the dist key column type doesn't match the const type such as int vs bigint.

So, I think I'll go ahead and defer pruning only when there is a parameter in the dist key

File: src/backend/distributed/planner/multi_router_planner.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/multi_router_planner.c]: I've done that, we can revert the commit `Defer pruning only for parameters` to see the previous implementation

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 3]
```diff
@@ -204,37 +261,109 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
...
+						errmsg("cannot perform distributed planning on this "
+							   "query because parameterized queries for SQL "
+							   "functions referencing distributed tables are "
+							   "not supported"),
+						errhint("Consider using PL/pgSQL functions instead.")));
+	}
 
-	/* modify tasks are always assigned using first-replica policy */
-	workerJob->taskList = FirstReplicaAssignTaskList(taskList);
+	shardIntervalList = list_make1(shardIntervalList);
```
Comment (JelteF) [inline src/backend/distributed/executor/citus_custom_scan.c]: This looked strange to me. It seems to me that it would be better if TargetShardIntervalForFastPathQuery already returned this. Looking at it's other usage we do the same there.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 4]
```diff
@@ -1703,6 +1720,39 @@ RouterJob(Query *originalQuery, PlannerRestrictionContext *plannerRestrictionCon
...
+	}
+
+	job->requiresMasterEvaluation = requiresMasterEvaluation;
+	return job;
+}
+
+
+/*
+ * SingleShardRouterTaskList is a wrapper around other corresponding task
+ * list generation functions speficit to single shard selects and modifications.
```
Comment (JelteF) [inline src/backend/distributed/planner/multi_router_planner.c]:
```suggestion
 * list generation functions speficic to single shard selects and modifications.
```

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 5]
```diff
@@ -175,6 +185,44 @@ CitusExecScan(CustomScanState *node)
...
+
+	/*
+	 * If we have not already copied the plan into this context, do it now.
+	 * Note that we could have copied the plan during CitusGenerateDeferredQueryStrings.
+	 */
+	if (GetMemoryChunkContext(distributedPlan) != CurrentMemoryContext)
+	{
+		distributedPlan = copyObject(distributedPlan);
+
+		scanState->distributedPlan = distributedPlan;
```
Comment (marcocitus) [inline src/backend/distributed/executor/citus_custom_scan.c]: I'm wondering whether we should only copy the Job and put it in the CitusScanState (e.g. jobToExecute). That way, the job in the DistributedPlan becomes more like a template of the Job at execution time.

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 6]
```diff
@@ -175,6 +185,44 @@ CitusExecScan(CustomScanState *node)
...
+	List *taskList = workerJob->taskList;
+
+	/* prevent concurrent placement changes */
+	AcquireMetadataLocks(taskList);
+
+	/*
+	 * We are taking locks on partitions of partitioned tables. These locks are
+	 * necessary for locking tables that appear in the SELECT part of the query.
+	 */
+	LockPartitionsInRelationList(distributedPlan->relationIdList, AccessShareLock);
```
Comment (marcocitus) [inline src/backend/distributed/executor/citus_custom_scan.c]: Do we need this here? Seems like we take locks in AdaptiveExecutor.

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 7]
```diff
@@ -204,37 +261,109 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
...
 		executorState->es_param_list_info = NULL;
+	}
 
-		if (workerJob->deferredPruning)
-		{
-			DeferredErrorMessage *planningError = NULL;
 
-			/* need to perform shard pruning, rebuild the task list from scratch */
-			taskList = RouterInsertTaskList(jobQuery, &planningError);
+	if (workerJob->requiresMasterEvaluation && !workerJob->deferredPruning)
```
[the remark this answers is not in the document]
Comment (marcocitus) [reply src/backend/distributed/executor/citus_custom_scan.c]: I don't follow. deferredPruning == true implies requiresMasterEvaluation == true, but deferredPruning == false does not imply requiresMasterEvaluation == true.

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 8]
```diff
@@ -188,12 +236,21 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
...
 
-	if (workerJob->requiresMasterEvaluation)
+	PlanState *planState = &(scanState->customScanState.ss.ps);
+	EState *executorState = planState->state;
+
+	/* citus only evaluates functions for modification queries */
+	bool modifyQueryRequiresMasterEvaluation =
+		workerJob->requiresMasterEvaluation && jobQuery->commandType != CMD_SELECT;
+
+	if (modifyQueryRequiresMasterEvaluation || workerJob->deferredPruning)
```
[the remark this answers is not in the document]
Comment (marcocitus) [reply src/backend/distributed/executor/citus_custom_scan.c]: deferredPruning can be removed from this if statement, since it's never true without modifyQueryRequiresMasterEvaluation being true.

[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: If that's an invariant the `workerJob` then that's not clear from the current code to me. Probably good to put an `Assert` at the top in that case.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: This function is only called when either `requiresMasterEvaluation` and/or `deferredPruning` is true. So it's not possible for both of them to be false.

Like I said in [this comment](https://github.com/citusdata/citus/pull/3369#discussion_r366858329) I think it's good to make that more clear by moving that check inside of the function.

File: src/backend/distributed/executor/citus_custom_scan.c [hunk 9]
```diff
@@ -129,11 +133,17 @@ CitusBeginScan(CustomScanState *node, EState *estate, int eflags)
 	ExecInitResultSlot(&scanState->customScanState.ss.ps, &TTSOpsMinimalTuple);
 #endif
 
-	distributedPlan = scanState->distributedPlan;
+	DistributedPlan *distributedPlan = scanState->distributedPlan;
+	Job *workerJob = distributedPlan->workerJob;
+	if (workerJob &&
+		(workerJob->requiresMasterEvaluation || workerJob->deferredPruning))
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: I actually find this easier to read in the sense that we get into string generation only when `workerJob->requiresMasterEvaluation || workerJob->deferredPruning`

The check for `workerJob ` being not null, that is unfortunate and very nice, I agree. That's because for INSERT .. SELECT via coordinator, Citus doesn't create a Job. 

Are you OK if I keep it like this?

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: It looks reasonable to remove this. We acquire the same locks in both AdaptiveExecutor and TaskTracker.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: > deferredPruning can be removed from this if statement, since it's never true without modifyQueryRequiresMasterEvaluation being true.

With this PR, this could happen now. For fast-path router queries with parameters, we only defer the pruning (e.g., not `modifyQueryRequiresMasterEvaluation `).  And, `ExecuteMasterEvaluableFunctions` handles both function evaluation and parameter evaluation.

From performance wise, it seems good to traverse the query once, not splitting the function evaluation from parameter evaluation. 

> The branching in this function is quite complex I think adding the following would help:

The branching in the function doesn't look too complicated to me. I think the missing part was comments as it is not clear to the readers why we're evaluating the function when `deferredPruning`.  How does it look now @JelteF ?

Additionally, I can consider renaming `ExecuteMasterEvaluableFunctions` to something that explains we evaluate parameters as well.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: I agree with Marco. @JelteF can you re-read the function now with the updated comments, let's see if this makes sense to you now?

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: Probably better to improve that with a separate PR?

File: src/backend/distributed/planner/multi_router_planner.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/multi_router_planner.c]: fyi: I generally don't prefer committing on the remote branch via github, as it complicates rebases and could potentially got lost. I added this change locally.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: Do you prefer for me to work on that in a separate PR on top of #3388?

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: I think it's okay to keep as it makes the intention clearer here indeed. But in that case there should at least be an `Assert` in the other function to ensure this invariant is not broken in the future.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: Comments definitely help indeed, but exiting early when possible makes the possible code paths much easier to understand. It also simplifies this check from effectively 3 boolean expressions to 2.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: Given `workerJob->requiresMasterEvaluation || workerJob->deferredPruning` (at call site of the function):
This:
```
workerJob->requiresMasterEvaluation && !workerJob->deferredPruning
```
is equal to this:
```
!workerJob->deferredPruning
```

[the remark this answers is not in the document]
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: It's also why the Assert below this block works

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 3 above)
Comment (JelteF) [reply src/backend/distributed/executor/citus_custom_scan.c]: unless you have to change tons of lines in that function I think it's fine to do it here, given that you also exported it in this PR.

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: I tried that, but didn't like much, I do want to get rid of anything related to commandType, it really shouldn't matter. The things that should matter are `workerJob->requiresMasterEvaluation` and `workerJob->deferredPruning`.

In the next release, I really want to get rid of `modifyQueryRequiresMasterEvaluation` just use `queryRequiresMasterEvaluation`

The current flow is not that complicated, we're calculating only two things:

1) Do we need to call `ExecuteMasterEvaluableFunctions`
2) Do we need to do pruning

I added an intermediate variable, and applied your below requested change

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: done

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 5 above)
Comment (marcocitus) [reply src/backend/distributed/executor/citus_custom_scan.c]: Maybe on top of #3388.

File: src/backend/distributed/planner/distributed_planner.c [hunk 10]
```diff
@@ -541,12 +541,25 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
...
+	}
+	else if (IsA(distributionKeyValue, Const))
+	{
+		planContext->plannerRestrictionContext->fastPathRestrictionContext->
+		distributionKeyValue = (Const *) distributionKeyValue;
+	}
+	else if (IsA(distributionKeyValue, Param))
+	{
+		planContext->plannerRestrictionContext->fastPathRestrictionContext->
+		distributionKeyHasParam = true;
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c]: Would be nice to unwrap this struct a little.

File: src/test/regress/expected/fast_path_router_modify.out [hunk 11]
```diff
@@ -371,21 +356,195 @@ PL/pgSQL function modify_fast_path_plpsql(integer,integer) line 3 at SQL stateme
...
+EXECUTE prepared_zero_shard_update(5);
+DEBUG:  Deferred pruning for a fast-path router query
+DEBUG:  Creating router plan
+DEBUG:  Plan is router executable
+EXECUTE prepared_zero_shard_update(6);
+DEBUG:  Deferred pruning for a fast-path router query
+DEBUG:  Creating router plan
+DEBUG:  Plan is router executable
+EXECUTE prepared_zero_shard_update(7);
+-- same test with fast-path disabled
```
Comment (marcocitus) [inline src/test/regress/expected/fast_path_router_modify.out]: Would be kind of nice to have a debug message here as well.

Comment (onderkalaci) [reply src/test/regress/expected/fast_path_router_modify.out]: done

File: src/backend/distributed/planner/distributed_planner.c (hunk 10 above)
Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c]: done

File: src/backend/distributed/executor/citus_custom_scan.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/executor/citus_custom_scan.c]: done
